### PR TITLE
feat: get_available_servers now accepts a filter

### PR DESCRIPTION
**Pull request by kylo252**
## Description

Allow `get_available_servers` to accept a query filter

The available keys are
- filetype (string): Only return servers with matching fileyupe

related: https://github.com/williamboman/mason.nvim/issues/86

### How has this been tested?

```lua
local all_servers = get_available_servers()
local js_servers = get_available_servers({ ft = "javascript" })
```

### TODOs
- [x] this current returns the full list of servers if the filetype query fails (regardless if the filetype is actually valid from the start)
- [ ] this should probably be used in `parse_packages_from_heuristics`
- [ ] we could probably use an `is_valid_server()`, see the test-case in https://github.com/williamboman/mason-lspconfig.nvim/commit/358f6279b5b1ba2378c8be2d00860403223e91be

### CODE REVIEW

**Comment by kylo252** [comment]
@williamboman, this one will be a bit more tricky, but I think it's quite useful, see https://github.com/williamboman/mason.nvim/issues/86

P.s. I've spent way too long trying to get the annotations for this in a reasonable state, but I'm not sure how you'd want it to be, so feel free to add any suggestions.

**Comment by williamboman** [comment]
@williambotman /apply

```diff
diff --git a/doc/mason-lspconfig.txt b/doc/mason-lspconfig.txt
index 30ab2d2..6762b84 100644
--- a/doc/mason-lspconfig.txt
+++ b/doc/mason-lspconfig.txt
@@ -246,10 +246,9 @@ get_available_servers({filter})
 
     Parameters: ~
         {filter}  (table|nil) A table with key-value pairs used to
-                  filter the list of server names. The available keys
-                  are:
-                  • filetype (string): Only return servers with matching
-                  fileyupe
+                  filter the list of server names. The available keys are:
+                  - filetype (string | string[]): Only return servers with
+                    matching filetype
 
     Returns: ~
         string[]
diff --git a/lua/mason-lspconfig/init.lua b/lua/mason-lspconfig/init.lua
index 667efd1..d6482b3 100644
--- a/lua/mason-lspconfig/init.lua
+++ b/lua/mason-lspconfig/init.lua
@@ -78,26 +78,53 @@ function M.get_installed_servers()
     end, registry.get_installed_package_names())
 end
 
+---@param filetype string | string[]
+local function is_server_in_filetype(filetype)
+    local filetype_mapping = require "mason-lspconfig.mappings.filetype"
+
+    local function get_servers_by_filetype(ft)
+        return filetype_mapping[ft] or {}
+    end
+
+    local server_candidates = _.compose(
+        _.set_of,
+        _.cond {
+            { _.is "string", get_servers_by_filetype },
+            { _.is "table", _.compose(_.flatten, _.map(get_servers_by_filetype)) },
+            { _.T, _.always {} },
+        }
+    )(filetype)
+
+    ---@param server_name string
+    ---@return boolean
+    return function(server_name)
+        return server_candidates[server_name]
+    end
+end
+
 ---Get a list of available servers in mason-registry
----@param filter {filetype: string}? (optional) used to filter the list of server names
---    The available keys are
---    - filetype (string): Only return servers with matching fileyupe
+---@param filter { filetype: string | string[] }?: (optional) Used to filter the list of server names.
+--- The available keys are
+---   - filetype (string | string[]): Only return servers with matching filetype
 ---@return string[]
 function M.get_available_servers(filter)
     local registry = require "mason-registry"
     local server_mapping = require "mason-lspconfig.mappings.server"
     local Optional = require "mason-core.optional"
-    local filetype_mapping = require "mason-lspconfig.mappings.filetype"
     filter = filter or {}
-    return Optional.of_nilable(filetype_mapping[filter.filetype])
-        :map(function(server_names)
-            return server_names
-        end)
-        :or_else_get(function()
-            return _.filter_map(function(pkg_name)
-                return Optional.of_nilable(server_mapping.package_to_lspconfig[pkg_name])
-            end, registry.get_all_package_names())
+    local predicates = {}
+
+    if filter.filetype then
+        table.insert(predicates, is_server_in_filetype(filter.filetype))
+    end
+
+    return _.filter_map(function(pkg_name)
+        return Optional.of_nilable(server_mapping.package_to_lspconfig[pkg_name]):map(function(server_name)
+            if #predicates == 0 or _.all_pass(predicates, server_name) then
+                return server_name
+            end
         end)
+    end, registry.get_all_package_names())
 end
 
 return M
diff --git a/tests/mason-lspconfig/api/api_spec.lua b/tests/mason-lspconfig/api/api_spec.lua
new file mode 100644
index 0000000..ec704ea
--- /dev/null
+++ b/tests/mason-lspconfig/api/api_spec.lua
@@ -0,0 +1,32 @@
+local mason_lspconfig = require "mason-lspconfig"
+local _ = require "mason-core.functional"
+
+describe("mason-lspconfig API", function()
+    it("should return all available servers", function()
+        assert.equals(116, #mason_lspconfig.get_available_servers())
+    end)
+
+    it("should return all available servers for given filetype", function()
+        assert.same(
+            { "terraformls", "tflint" },
+            _.sort_by(
+                _.identity,
+                mason_lspconfig.get_available_servers {
+                    filetype = "terraform",
+                }
+            )
+        )
+    end)
+
+    it("should return all available servers for given filetypes", function()
+        assert.same(
+            { "lemminx", "taplo" },
+            _.sort_by(
+                _.identity,
+                mason_lspconfig.get_available_servers {
+                    filetype = { "xml", "xsd", "xsl", "toml" },
+                }
+            )
+        )
+    end)
+end)
```

**Comment by williamboman** [comment]
@williambotman /apply

```diff
diff --git a/doc/mason-lspconfig.txt b/doc/mason-lspconfig.txt
index 30ab2d2..6762b84 100644
--- a/doc/mason-lspconfig.txt
+++ b/doc/mason-lspconfig.txt
@@ -246,10 +246,9 @@ get_available_servers({filter})
 
     Parameters: ~
         {filter}  (table|nil) A table with key-value pairs used to
-                  filter the list of server names. The available keys
-                  are:
-                  • filetype (string): Only return servers with matching
-                  fileyupe
+                  filter the list of server names. The available keys are:
+                  - filetype (string | string[]): Only return servers with
+                    matching filetype
 
     Returns: ~
         string[]
diff --git a/lua/mason-lspconfig/init.lua b/lua/mason-lspconfig/init.lua
index 667efd1..d6482b3 100644
--- a/lua/mason-lspconfig/init.lua
+++ b/lua/mason-lspconfig/init.lua
@@ -78,26 +78,53 @@ function M.get_installed_servers()
     end, registry.get_installed_package_names())
 end
 
+---@param filetype string | string[]
+local function is_server_in_filetype(filetype)
+    local filetype_mapping = require "mason-lspconfig.mappings.filetype"
+
+    local function get_servers_by_filetype(ft)
+        return filetype_mapping[ft] or {}
+    end
+
+    local server_candidates = _.compose(
+        _.set_of,
+        _.cond {
+            { _.is "string", get_servers_by_filetype },
+            { _.is "table", _.compose(_.flatten, _.map(get_servers_by_filetype)) },
+            { _.T, _.always {} },
+        }
+    )(filetype)
+
+    ---@param server_name string
+    ---@return boolean
+    return function(server_name)
+        return server_candidates[server_name]
+    end
+end
+
 ---Get a list of available servers in mason-registry
----@param filter {filetype: string}? (optional) used to filter the list of server names
---    The available keys are
---    - filetype (string): Only return servers with matching fileyupe
+---@param filter { filetype: string | string[] }?: (optional) Used to filter the list of server names.
+--- The available keys are
+---   - filetype (string | string[]): Only return servers with matching filetype
 ---@return string[]
 function M.get_available_servers(filter)
     local registry = require "mason-registry"
     local server_mapping = require "mason-lspconfig.mappings.server"
     local Optional = require "mason-core.optional"
-    local filetype_mapping = require "mason-lspconfig.mappings.filetype"
     filter = filter or {}
-    return Optional.of_nilable(filetype_mapping[filter.filetype])
-        :map(function(server_names)
-            return server_names
-        end)
-        :or_else_get(function()
-            return _.filter_map(function(pkg_name)
-                return Optional.of_nilable(server_mapping.package_to_lspconfig[pkg_name])
-            end, registry.get_all_package_names())
+    local predicates = {}
+
+    if filter.filetype then
+        table.insert(predicates, is_server_in_filetype(filter.filetype))
+    end
+
+    return _.filter_map(function(pkg_name)
+        return Optional.of_nilable(server_mapping.package_to_lspconfig[pkg_name]):map(function(server_name)
+            if #predicates == 0 or _.all_pass(predicates, server_name) then
+                return server_name
+            end
         end)
+    end, registry.get_all_package_names())
 end
 
 return M
diff --git a/tests/mason-lspconfig/api/api_spec.lua b/tests/mason-lspconfig/api/api_spec.lua
new file mode 100644
index 0000000..ec704ea
--- /dev/null
+++ b/tests/mason-lspconfig/api/api_spec.lua
@@ -0,0 +1,32 @@
+local mason_lspconfig = require "mason-lspconfig"
+local _ = require "mason-core.functional"
+
+describe("mason-lspconfig API", function()
+    it("should return all available servers", function()
+        assert.equals(116, #mason_lspconfig.get_available_servers())
+    end)
+
+    it("should return all available servers for given filetype", function()
+        assert.same(
+            { "terraformls", "tflint" },
+            _.sort_by(
+                _.identity,
+                mason_lspconfig.get_available_servers {
+                    filetype = "terraform",
+                }
+            )
+        )
+    end)
+
+    it("should return all available servers for given filetypes", function()
+        assert.same(
+            { "lemminx", "taplo" },
+            _.sort_by(
+                _.identity,
+                mason_lspconfig.get_available_servers {
+                    filetype = { "xml", "xsd", "xsl", "toml" },
+                }
+            )
+        )
+    end)
+end)
```

**Comment by williamboman** [comment]
> P.s. I've spent way too long trying to get the annotations for this in a reasonable state, but I'm not sure how you'd want it to be, so feel free to add any suggestions.

They look good to me! I changed the signature to allow both a string as well as a list of strings, felt more natural to me to allow both variants.

>  this should probably be used in parse_packages_from_heuristics

Good catch! Should be easy to incorporate, will look into it

**Comment by williamboman** [comment]
@williambotman /apply

```diff
diff --git a/tests/mason-lspconfig/api/api_spec.lua b/tests/mason-lspconfig/api/api_spec.lua
index ec704ea..758263d 100644
--- a/tests/mason-lspconfig/api/api_spec.lua
+++ b/tests/mason-lspconfig/api/api_spec.lua
@@ -29,4 +29,13 @@ describe("mason-lspconfig API", function()
             )
         )
     end)
+
+    it("should return no servers if filetype predicate has no matches", function()
+        assert.same(
+            {},
+            mason_lspconfig.get_available_servers {
+                filetype = { "thisfiletypesimplydoesntexist" },
+            }
+        )
+    end)
 end)
```

**Review by williamboman — APPROVED** [review]
LGTM! Let's look into the TODOs in later PRs. I'll wait with merge until I have your blessing @kylo252, not sure if I messed something up with my changes

File: tests/mason-lspconfig/api/api_spec.lua
Comment (kylo252) [inline tests/mason-lspconfig/api/api_spec.lua:44]: I stumbled upon this test-case while trying a few things, but I'm not sure what the behavior should be.

do you think it's worth it? feels like it would be useful for the CI at least, or is this out-of-scope for mason-lspconfig itself?

File: tests/mason-lspconfig/api/api_spec.lua
Comment (kylo252) [inline tests/mason-lspconfig/api/api_spec.lua:8]: I figured it's easier to not hard-code the number of servers, but feel to revert it if you don't want to hard-code the dependency itself :)

File: tests/mason-lspconfig/api/api_spec.lua
Comment (williamboman) [inline tests/mason-lspconfig/api/api_spec.lua:52]: Hmm this doesn't seem like intended behaviour to me. The test fixtures are super hacky, I just shoehorned them in for some specific test cases.

I'd expect this to return `{ "dummylsp", "dummy2lsp" }`?  Mason doesn't support registering custom packages yet, so this is in many ways undefined behaviour, there's no right or wrong

Comment (kylo252) [reply tests/mason-lspconfig/api/api_spec.lua:52]: > I'd expect this to return `{ "dummylsp", "dummy2lsp" }`? Mason doesn't support registering custom packages yet, so this is in many ways undefined behaviour, there's no right or wrong

yeah, that's what I had expected as well. I'll add another item to the TODO list for traceability in later PRs.

Comment (kylo252) [reply tests/mason-lspconfig/api/api_spec.lua]: @williamboman, anything else before merging this?

**Pull request merged**
